crates/net/discv5/src/lib.rs

```diff
@@ -0,0 +1,789 @@
+            node_2_node_id=format!("{:#}", node_2_enr.node_id()),
+            "started nodes"
+        );
+
+        // test
+
+        // add node_2 to discovery handle of node_1 (should add node to discv5 kbuckets)
+        let node_2_enr_reth_compatible_ty: Enr<SecretKey> =
+            EnrCombinedKeyWrapper(node_2_enr.clone()).into();
+        node_1.add_node_to_routing_table(node_2_enr_reth_compatible_ty).unwrap();
+
+        // verify node_2 is in KBuckets of node_1:discv5
+        assert!(
+            node_1.with_discv5(|discv5| discv5.table_entries_id().contains(&node_2_enr.node_id()))
+        );
+
+        // manually trigger connection from node_1 to node_2
+        node_1.with_discv5(|discv5| discv5.send_ping(node_2_enr.clone())).await.unwrap();
+
+        // verify node_1:discv5 is connected to node_2:discv5 and vv
+        let event_2_v5 = stream_2.recv().await.unwrap();
+        let event_1_v5 = stream_1.recv().await.unwrap();
+        matches!(
+            event_1_v5,
+            discv5::Event::SessionEstablished(node, socket) if node == node_2_enr && socket == node_2_enr.udp4_socket().unwrap().into()
+        );
+        matches!(
+            event_2_v5,
+            discv5::Event::SessionEstablished(node, socket) if node == node_1_enr && socket == node_1_enr.udp4_socket().unwrap().into()
+        );
+
+        // verify node_1 is in KBuckets of node_2:discv5
+        let event_2_v5 = stream_2.recv().await.unwrap();
+        matches!(
+            event_2_v5,
+            discv5::Event::NodeInserted { node_id, replaced } if node_id == node_1_enr.node_id() && replaced.is_none()
+        );
+    }
+
+    #[test]
+    fn discovered_enr_disc_socket_missing() {
+        reth_tracing::init_test_tracing();
+
+        // rig test
+        const REMOTE_RLPX_PORT: u16 = 30303;
+        let remote_socket = "104.28.44.25:9000".parse().unwrap();
+        let remote_key = CombinedKey::generate_secp256k1();
+        let remote_enr = Enr::builder().tcp4(REMOTE_RLPX_PORT).build(&remote_key).unwrap();
+
+        let mut discv5 = discv5_noop();
+
+        // test
+        let filtered_peer = discv5.on_discovered_peer(&remote_enr, remote_socket);
+
+        assert_eq!(
+            NodeRecord {
+                address: remote_socket.ip(),
+                udp_port: remote_socket.port(),
+                tcp_port: REMOTE_RLPX_PORT,
+                id: enr_to_discv4_id(&remote_enr),
+            },
+            filtered_peer.unwrap().node_record
+        )
+    }
+
+    // Copied from sigp/discv5
+    #[allow(unreachable_pub)]
+    #[allow(unused)]
+    #[allow(clippy::assign_op_pattern)]
+    mod sigp {
+        use enr::{
+            k256::sha2::digest::generic_array::{typenum::U32, GenericArray},
+            NodeId,
+        };
+        use uint::construct_uint;
+
+        construct_uint! {
+            /// 256-bit unsigned integer.
+            pub(super) struct U256(4);
+        }
```

Comment:
can we use existing U256?

Reply:
I just wanted to copy paste the code exactly as it is in sigp, makes easier to follow along than if its some adaptation

Reply:
we already have an u256 type